File: src/scripts/create_token.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'base64'
require 'fileutils'
require 'json'
require 'net/http'
require 'openssl'
require 'time'
require 'uri'

module GitHubAppToken
  RUNTIME_PREFIX = 'GITHUB_APP_TOKEN_'
  ENV_NAME_PATTERN = /\A[A-Za-z_][A-Za-z0-9_]*\z/
  INSTALLATION_LOOKUPS = %w[repo org user app-list none].freeze

  class InputError < StandardError; end
  class GitHubApiError < StandardError; end

  class GitHubClient
    def initialize(api_url:, api_version:)
      @api_url = api_url
      @api_version = api_version
    end

    def request(method, request_path, auth_token:, body: nil)
      uri = URI(api_url_with_path(request_path))
      request = build_request(method, uri, auth_token, body)
      response = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(request)
      end

      data = GitHubAppToken.parse_json_body(response.body)
      return data if response.is_a?(Net::HTTPSuccess)

      detail = data['message'] ? ": #{data['message']}" : ''
      raise GitHubApiError, "GitHub API #{method} #{request_path} failed with #{response.code}#{detail}"
    end

    private

    def api_url_with_path(request_path)
      "#{@api_url.sub(%r{/+\z}, '')}/#{request_path.sub(%r{\A/+}, '')}"
    end

    def build_request(method, uri, auth_token, body)
      klass = {
        'GET' => Net::HTTP::Get,
        'POST' => Net::HTTP::Post
      }.fetch(method) { raise InputError, "Unsupported GitHub API method: #{method}" }

      request = klass.new(uri)
      request['Accept'] = 'application/vnd.github+json'
      request['Authorization'] = "Bearer #{auth_token}"
      request['Content-Type'] = 'application/json'
      request['X-GitHub-Api-Version'] = @api_version
      request.body = JSON.generate(body) unless body.nil?
      request
    end
  end

  module_function

  def parse_boolean(value, default_value = false)
    return default_value if value.nil? || value == ''
    return value if value == true || value == false

    normalized = value.to_s.strip.downcase
    return true if %w[true 1 yes y].include?(normalized)
    return false if %w[false 0 no n].include?(normalized)

    raise InputError, "Invalid boolean value: #{value}"
  end

  def expand_env_reference(value, env = ENV)
    return value unless value.is_a?(String)

    value.gsub(/\$\{([A-Za-z_][A-Za-z0-9_]*)\}|\$([A-Za-z_][A-Za-z0-9_]*)/) do
      env[Regexp.last_match(1) || Regexp.last_match(2)] || ''
    end
  end

  def runtime_value(env, name, fallback = '')
    expand_env_reference(env.fetch("#{RUNTIME_PREFIX}#{name}", fallback), env)
  end

  def read_runtime_options(env = ENV)
    {
      app_id: runtime_value(env, 'APP_ID'),
      app_id_env_var: runtime_value(env, 'APP_ID_ENV_VAR', 'GITHUB_APP_ID'),
      private_key_env_var: runtime_value(env, 'PRIVATE_KEY_ENV_VAR', 'GITHUB_APP_PRIVATE_KEY'),
      private_key_base64: parse_boolean(runtime_value(env, 'PRIVATE_KEY_BASE64', 'false')),
      installation_id: runtime_value(env, 'INSTALLATION_ID'),
      installation_id_env_var: runtime_value(env, 'INSTALLATION_ID_ENV_VAR', 'GITHUB_APP_INSTALLATION_ID'),
      owner: runtime_value(env, 'OWNER', '${CIRCLE_PROJECT_USERNAME}'),
      repo: runtime_value(env, 'REPO', '${CIRCLE_PROJECT_REPONAME}'),
      installation_lookup: runtime_value(env, 'INSTALLATION_LOOKUP', 'repo'),
      repositories: runtime_value(env, 'REPOSITORIES'),
      repository_ids: runtime_value(env, 'REPOSITORY_IDS'),
      permissions: runtime_value(env, 'PERMISSIONS'),
      github_api_url: runtime_value(env, 'GITHUB_API_URL', 'https://api.github.com'),
      github_api_version: runtime_value(env, 'GITHUB_API_VERSION', '2026-03-10'),
      token_env_var: runtime_value(env, 'TOKEN_ENV_VAR', 'GH_APP_TOKEN'),
      export_env_prefix: runtime_value(env, 'EXPORT_ENV_PREFIX', 'GH_APP'),
      output_json: runtime_value(env, 'OUTPUT_JSON', '/tmp/github-app-token-result.json'),
      include_token_in_output_json: parse_boolean(runtime_value(env, 'INCLUDE_TOKEN_IN_OUTPUT_JSON', 'false'))
    }
  end

  def normalize_private_key(raw_value, base64: false)
    raise InputError, 'GitHub App private key is empty.' if raw_value.to_s.strip.empty?

    private_key = raw_value.to_s.strip
    private_key = Base64.decode64(private_key.gsub(/\s+/, '')).strip if base64
    private_key = private_key.gsub('\\n', "\n").strip

    has_header = private_key.match?(/-----BEGIN (?:RSA )?PRIVATE KEY-----/)
    has_footer = private_key.match?(/-----END (?:RSA )?PRIVATE KEY-----/)
    raise InputError, 'GitHub App private key must be a PEM private key.' unless has_header && has_footer

    private_key.end_with?("\n") ? private_key : "#{private_key}\n"
  end

  def validate_env_name(value, label)
    return if value.to_s.match?(ENV_NAME_PATTERN)

    raise InputError, "#{label} must be a valid environment variable name."
  end

  def clean(value)
    value.to_s.strip
  end

  def parse_comma_list(value)
    clean(value).empty? ? [] : clean(value).split(',').map(&:strip).reject(&:empty?)
  end

  def parse_repository_ids(value)
    parse_comma_list(value).map do |item|
      raise InputError, 'repository_ids must contain only numeric IDs.' unless item.match?(/\A\d+\z/)

      item.to_i
    end
  end

  def parse_permissions(value)
    return nil if clean(value).empty?

    parsed = JSON.parse(value)
    raise InputError, 'permissions must be a JSON object.' unless parsed.is_a?(Hash)

    parsed
  rescue JSON::ParserError => e
    raise InputError, "permissions must be a valid JSON object: #{e.message}"
  end

  def create_app_jwt(app_id:, private_key:, now: Time.now)
    now_seconds = now.to_i
    header = base64url(JSON.generate({ alg: 'RS256', typ: 'JWT' }))
    payload = base64url(JSON.generate({ iat: now_seconds - 60, exp: now_seconds + 540, iss: app_id }))
    body = "#{header}.#{payload}"
    key = OpenSSL::PKey.read(private_key)
    signature = base64url(key.sign(OpenSSL::Digest.new('SHA256'), body))
    "#{body}.#{signature}"
  end

  def lookup_installation_id(options:, jwt:, client:)
    owner = clean(options[:owner])
    repo = clean(options[:repo])

    case options[:installation_lookup]
    when 'repo'
      raise InputError, 'owner and repo are required for repo installation lookup.' if owner.empty? || repo.empty?

      data = client.request('GET', "/repos/#{encode_path_part(owner)}/#{encode_path_part(repo)}/installation",
                            auth_token: jwt)
      clean(data['id'])
    when 'org', 'user'
      raise InputError, "owner is required for #{options[:installation_lookup]} installation lookup." if owner.empty?

      segment = options[:installation_lookup] == 'org' ? 'orgs' : 'users'
      data = client.request('GET', "/#{segment}/#{encode_path_part(owner)}/installation", auth_token: jwt)
      clean(data['id'])
    when 'app-list'
      raise InputError, 'owner is required for app-list installation lookup.' if owner.empty?

      data = client.request('GET', '/app/installations', auth_token: jwt)
      installations = data.is_a?(Array) ? data : data.fetch('installations', [])
      match = installations.find do |installation|
        installation.dig('account', 'login').to_s.downcase == owner.downcase
      end
      raise InputError, "No GitHub App installation found for #{owner}." unless match

      clean(match['id'])
    else
      raise InputError, 'installation_id is required when installation_lookup is none.'
    end
  end

  def build_access_token_body(repositories:, repository_ids:, permissions:)
    body = {}
    body[:repositories] = repositories unless repositories.empty?
    body[:repository_ids] = repository_ids unless repository_ids.empty?
    body[:permissions] = permissions unless permissions.nil?
    body.empty? ? nil : body
  end

  def create_token(options, env: ENV, client: nil, now: Time.now)
    validate_options(options)

    app_id = clean(options[:app_id])
    app_id = clean(env[options[:app_id_env_var]]) if app_id.empty?
    raise InputError, 'GitHub App ID is required through app_id or app_id_env_var.' if app_id.empty?

    private_key = normalize_private_key(env[options[:private_key_env_var]],
                                        base64: options[:private_key_base64])
    explicit_installation_id = clean(options[:installation_id])
    explicit_installation_id = clean(env[options[:installation_id_env_var]]) if explicit_installation_id.empty?
    repositories = parse_comma_list(options[:repositories])
    repository_ids = parse_repository_ids(options[:repository_ids])
    permissions = parse_permissions(options[:permissions])

    unless repositories.empty? || repository_ids.empty?
      raise InputError, 'repositories and repository_ids cannot be used together.'
    end

    client ||= GitHubClient.new(api_url: options[:github_api_url], api_version: options[:github_api_version])
    jwt = create_app_jwt(app_id: app_id, private_key: private_key, now: now)
    installation_id = explicit_installation_id
    installation_id = lookup_installation_id(options: options, jwt: jwt, client: client) if installation_id.empty?

    token_response = client.request(
      'POST',
      "/app/installations/#{encode_path_part(installation_id)}/access_tokens",
      auth_token: jwt,
      body: build_access_token_body(repositories: repositories, repository_ids: repository_ids,
                                    permissions: permissions)
    )
    raise GitHubApiError, 'GitHub access token response did not include a token.' if token_response['token'].to_s.empty?

    metadata = build_metadata(token_response, installation_id, repositories, options[:token_env_var])
    metadata[:token] = token_response['token'] if options[:include_token_in_output_json]

    write_output_json(options[:output_json], metadata)
    write_bash_env(env['BASH_ENV'], options[:token_env_var], options[:export_env_prefix], options[:output_json],
                   token_response, installation_id)
    metadata
  end

  def redact(text, secrets)
    secrets.compact.uniq.select { |secret| secret.is_a?(String) && secret.length >= 4 }.sort_by { |s| -s.length }
           .reduce(text.to_s) { |message, secret| message.gsub(secret, '[REDACTED]') }
  end

  def parse_json_body(body)
    return {} if body.to_s.empty?

    JSON.parse(body)
  end

  def validate_options(options)
    validate_env_name(options[:app_id_env_var], 'app_id_env_var')
    validate_env_name(options[:private_key_env_var], 'private_key_env_var')
    validate_env_name(options[:installation_id_env_var], 'installation_id_env_var')
    validate_env_name(options[:token_env_var], 'token_env_var')
    validate_env_name(options[:export_env_prefix], 'export_env_prefix')

    return if INSTALLATION_LOOKUPS.include?(options[:installation_lookup])

    raise InputError, 'installation_lookup must be repo, org, user, app-list, or none.'
  end

  def build_metadata(token_response, installation_id, requested_repositories, token_env_var)
    repositories = if token_response['repositories'].is_a?(Array)
                     token_response['repositories'].filter_map { |repository| repository['name'] }
                   else
                     requested_repositories
                   end

    {
      installation_id: integer_or_string(installation_id),
      expires_at: token_response['expires_at'],
      repository_selection: token_response['repository_selection'],
      permissions: token_response['permissions'] || {},
      repositories: repositories,
      token_exported_to: token_env_var
    }
  end

  def write_output_json(output_json, metadata)
    FileUtils.mkdir_p(File.dirname(output_json))
    File.write(output_json, "#{JSON.pretty_generate(metadata)}\n", mode: 'w', perm: 0o600)
    File.chmod(0o600, output_json)
  end

  def write_bash_env(bash_env_path, token_env_var, export_env_prefix, output_json, token_response, installation_id)
    raise InputError, 'BASH_ENV is not set; CircleCI needs it to export values for later steps.' if bash_env_path.to_s.empty?

    FileUtils.mkdir_p(File.dirname(bash_env_path))
    lines = [
      "export #{token_env_var}=#{shell_quote(token_response['token'])}",
      "export #{export_env_prefix}_INSTALLATION_ID=#{shell_quote(installation_id)}",
      "export #{export_env_prefix}_EXPIRES_AT=#{shell_quote(token_response['expires_at'])}",
      "export #{export_env_prefix}_OUTPUT_JSON=#{shell_quote(output_json)}"
    ]
    File.open(bash_env_path, File::WRONLY | File::APPEND | File::CREAT, 0o600) do |file|
      file.write("#{lines.join("\n")}\n")
    end
  end

  def shell_quote(value)
    "'#{value.to_s.gsub("'", "'\\\\''")}'"
  end

  def base64url(value)
    Base64.urlsafe_encode64(value, padding: false)
  end

  def encode_path_part(value)
    URI.encode_www_form_component(value.to_s).gsub('%2F', '/')
  end

  def integer_or_string(value)
    value.to_s.match?(/\A\d+\z/) ? value.to_i : value
  end
end

if __FILE__ == $PROGRAM_NAME
  options = nil
  begin
    options = GitHubAppToken.read_runtime_options(ENV)
    metadata = GitHubAppToken.create_token(options)
    puts "GitHub App installation token exported to #{metadata[:token_exported_to]}; " \
         "metadata written to #{options[:output_json]}."
  rescue StandardError => e
    private_key = options && ENV[options[:private_key_env_var]]
    warn GitHubAppToken.redact("#{e.class}: #{e.message}\n#{e.backtrace&.join("\n")}", [private_key])
    exit 1
  end
end

File: src/scripts/create_token.mjs
#!/usr/bin/env node
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

const RUNTIME_PREFIX = 'GITHUB_APP_TOKEN_';
const ENV_NAME_PATTERN = /^[A-Za-z_][A-Za-z0-9_]*$/;
const INSTALLATION_LOOKUPS = new Set(['repo', 'org', 'user', 'app-list', 'none']);

export class InputError extends Error {
  constructor(message) {
    super(message);
    this.name = 'InputError';
  }
}

export class GitHubApiError extends Error {
  constructor(message) {
    super(message);
    this.name = 'GitHubApiError';
  }
}

export const parseBoolean = (value, defaultValue = false) => {
  if (value === undefined || value === null || value === '') {
    return defaultValue;
  }

  if (typeof value === 'boolean') {
    return value;
  }

  const normalized = String(value).trim().toLowerCase();
  if (['true', '1', 'yes', 'y'].includes(normalized)) {
    return true;
  }
  if (['false', '0', 'no', 'n'].includes(normalized)) {
    return false;
  }

  throw new InputError(`Invalid boolean value: ${value}`);
};

export const expandEnvReference = (value, env = process.env) => {
  if (typeof value !== 'string') {
    return value;
  }

  const trimmed = value.trim();
  const braced = trimmed.match(/^\$\{([A-Za-z_][A-Za-z0-9_]*)\}$/);
  if (braced) {
    return env[braced[1]] ?? '';
  }

  const plain = trimmed.match(/^\$([A-Za-z_][A-Za-z0-9_]*)$/);
  if (plain) {
    return env[plain[1]] ?? '';
  }

  return value;
};

const runtimeValue = (env, name, fallback = '') =>
  expandEnvReference(env[`${RUNTIME_PREFIX}${name}`] ?? fallback, env);

export const readRuntimeOptions = (env = process.env) => ({
  appId: runtimeValue(env, 'APP_ID'),
  appIdEnvVar: runtimeValue(env, 'APP_ID_ENV_VAR', 'GITHUB_APP_ID'),
  privateKeyEnvVar: runtimeValue(env, 'PRIVATE_KEY_ENV_VAR', 'GITHUB_APP_PRIVATE_KEY'),
  privateKeyBase64: parseBoolean(runtimeValue(env, 'PRIVATE_KEY_BASE64', 'false')),
  installationId: runtimeValue(env, 'INSTALLATION_ID'),
  installationIdEnvVar: runtimeValue(env, 'INSTALLATION_ID_ENV_VAR', 'GITHUB_APP_INSTALLATION_ID'),
  owner: runtimeValue(env, 'OWNER', '${CIRCLE_PROJECT_USERNAME}'),
  repo: runtimeValue(env, 'REPO', '${CIRCLE_PROJECT_REPONAME}'),
  installationLookup: runtimeValue(env, 'INSTALLATION_LOOKUP', 'repo'),
  repositories: runtimeValue(env, 'REPOSITORIES'),
  repositoryIds: runtimeValue(env, 'REPOSITORY_IDS'),
  permissions: runtimeValue(env, 'PERMISSIONS'),
  githubApiUrl: runtimeValue(env, 'GITHUB_API_URL', 'https://api.github.com'),
  githubApiVersion: runtimeValue(env, 'GITHUB_API_VERSION', '2026-03-10'),
  tokenEnvVar: runtimeValue(env, 'TOKEN_ENV_VAR', 'GH_APP_TOKEN'),
  exportEnvPrefix: runtimeValue(env, 'EXPORT_ENV_PREFIX', 'GH_APP'),
  outputJson: runtimeValue(env, 'OUTPUT_JSON', '/tmp/github-app-token-result.json'),
  includeTokenInOutputJson: parseBoolean(runtimeValue(env, 'INCLUDE_TOKEN_IN_OUTPUT_JSON', 'false')),
});

export const normalizePrivateKey = (rawValue, { base64 = false } = {}) => {
  if (!rawValue || !String(rawValue).trim()) {
    throw new InputError('GitHub App private key is empty.');
  }

  let privateKey = String(rawValue).trim();
  if (base64) {
    privateKey = Buffer.from(privateKey.replace(/\s+/g, ''), 'base64').toString('utf8').trim();
  }

  privateKey = privateKey.replace(/\\n/g, '\n').trim();
  const hasHeader = /-----BEGIN (?:RSA )?PRIVATE KEY-----/.test(privateKey);
  const hasFooter = /-----END (?:RSA )?PRIVATE KEY-----/.test(privateKey);
  if (!hasHeader || !hasFooter) {
    throw new InputError('GitHub App private key must be a PEM private key.');
  }

  return privateKey.endsWith('\n') ? privateKey : `${privateKey}\n`;
};

const validateEnvName = (value, label) => {
  if (!ENV_NAME_PATTERN.test(value)) {
    throw new InputError(`${label} must be a valid environment variable name.`);
  }
};

const clean = (value) => String(value ?? '').trim();

export const parseCommaList = (value) =>
  clean(value)
    ? clean(value)
        .split(',')
        .map((item) => item.trim())
        .filter(Boolean)
    : [];

export const parseRepositoryIds = (value) =>
  parseCommaList(value).map((item) => {
    if (!/^\d+$/.test(item)) {
      throw new InputError('repository_ids must contain only numeric IDs.');
    }
    return Number(item);
  });

export const parsePermissions = (value) => {
  if (!clean(value)) {
    return undefined;
  }

  let parsed;
  try {
    parsed = JSON.parse(value);
  } catch (error) {
    throw new InputError(`permissions must be a valid JSON object: ${error.message}`);
  }

  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new InputError('permissions must be a JSON object.');
  }

  return parsed;
};

export const createAppJwt = ({ appId, privateKey, now = new Date() }) => {
  const nowSeconds = Math.floor(now.getTime() / 1000);
  const header = Buffer.from(JSON.stringify({ alg: 'RS256', typ: 'JWT' })).toString('base64url');
  const payload = Buffer.from(
    JSON.stringify({
      iat: nowSeconds - 60,
      exp: nowSeconds + 540,
      iss: appId,
    }),
  ).toString('base64url');
  const body = `${header}.${payload}`;
  const signer = crypto.createSign('RSA-SHA256');
  signer.update(body);
  signer.end();
  const signature = signer.sign(crypto.createPrivateKey(privateKey)).toString('base64url');
  return `${body}.${signature}`;
};

const apiUrlWithPath = (apiUrl, requestPath) =>
  `${apiUrl.replace(/\/+$/, '')}${requestPath.startsWith('/') ? requestPath : `/${requestPath}`}`;

export const requestGitHub = async ({
  apiUrl,
  apiVersion,
  method,
  requestPath,
  authToken,
  body,
  fetchImpl = globalThis.fetch,
}) => {
  const response = await fetchImpl(apiUrlWithPath(apiUrl, requestPath), {
    method,
    headers: {
      Accept: 'application/vnd.github+json',
      Authorization: `Bearer ${authToken}`,
      'Content-Type': 'application/json',
      'X-GitHub-Api-Version': apiVersion,
    },
    body: body === undefined ? undefined : JSON.stringify(body),
  });

  const text = await response.text();
  const data = text ? JSON.parse(text) : {};
  if (!response.ok) {
    const detail = data.message ? `: ${data.message}` : '';
    throw new GitHubApiError(`GitHub API ${method} ${requestPath} failed with ${response.status}${detail}`);
  }

  return data;
};

const encodePathPart = (value) => encodeURIComponent(value).replace(/%2F/g, '/');

export const lookupInstallationId = async ({ options, jwt, fetchImpl }) => {
  const owner = clean(options.owner);
  const repo = clean(options.repo);

  if (options.installationLookup === 'repo') {
    if (!owner || !repo) {
      throw new InputError('owner and repo are required for repo installation lookup.');
    }
    const data = await requestGitHub({
      apiUrl: options.githubApiUrl,
      apiVersion: options.githubApiVersion,
      method: 'GET',
      requestPath: `/repos/${encodePathPart(owner)}/${encodePathPart(repo)}/installation`,
      authToken: jwt,
      fetchImpl,
    });
    return clean(data.id);
  }

  if (options.installationLookup === 'org' || options.installationLookup === 'user') {
    if (!owner) {
      throw new InputError(`owner is required for ${options.installationLookup} installation lookup.`);
    }
    const segment = options.installationLookup === 'org' ? 'orgs' : 'users';
    const data = await requestGitHub({
      apiUrl: options.githubApiUrl,
      apiVersion: options.githubApiVersion,
      method: 'GET',
      requestPath: `/${segment}/${encodePathPart(owner)}/installation`,
      authToken: jwt,
      fetchImpl,
    });
    return clean(data.id);
  }

  if (options.installationLookup === 'app-list') {
    if (!owner) {
      throw new InputError('owner is required for app-list installation lookup.');
    }
    const data = await requestGitHub({
      apiUrl: options.githubApiUrl,
      apiVersion: options.githubApiVersion,
      method: 'GET',
      requestPath: '/app/installations',
      authToken: jwt,
      fetchImpl,
    });
    const installations = Array.isArray(data) ? data : data.installations ?? [];
    const match = installations.find((installation) => {
      const login = installation?.account?.login ?? '';
      return login.toLowerCase() === owner.toLowerCase();
    });
    if (!match) {
      throw new InputError(`No GitHub App installation found for ${owner}.`);
    }
    return clean(match.id);
  }

  throw new InputError('installation_id is required when installation_lookup is none.');
};

export const buildAccessTokenBody = ({ repositories, repositoryIds, permissions }) => {
  const body = {};
  if (repositories.length > 0) {
    body.repositories = repositories;
  }
  if (repositoryIds.length > 0) {
    body.repository_ids = repositoryIds;
  }
  if (permissions) {
    body.permissions = permissions;
  }
  return Object.keys(body).length > 0 ? body : undefined;
};

const shellQuote = (value) => `'${String(value).replace(/'/g, `'\\''`)}'`;

const writeBashEnv = ({ bashEnvPath, tokenEnvVar, exportEnvPrefix, outputJson, result }) => {
  if (!bashEnvPath) {
    throw new InputError('BASH_ENV is not set; CircleCI needs it to export values for later steps.');
  }

  fs.mkdirSync(path.dirname(bashEnvPath), { recursive: true });
  const lines = [
    `export ${tokenEnvVar}=${shellQuote(result.token)}`,
    `export ${exportEnvPrefix}_INSTALLATION_ID=${shellQuote(result.installationId)}`,
    `export ${exportEnvPrefix}_EXPIRES_AT=${shellQuote(result.expiresAt)}`,
    `export ${exportEnvPrefix}_OUTPUT_JSON=${shellQuote(outputJson)}`,
  ];
  fs.appendFileSync(bashEnvPath, `${lines.join('\n')}\n`, { mode: 0o600 });
};

const writeOutputJson = ({ outputJson, metadata }) => {
  fs.mkdirSync(path.dirname(outputJson), { recursive: true });
  fs.writeFileSync(outputJson, `${JSON.stringify(metadata, null, 2)}\n`, { mode: 0o600 });
  fs.chmodSync(outputJson, 0o600);
};

export const redact = (text, secrets) => {
  let redacted = String(text);
  const uniqueSecrets = [...new Set(secrets.filter((secret) => typeof secret === 'string' && secret.length >= 4))];
  for (const secret of uniqueSecrets.sort((a, b) => b.length - a.length)) {
    redacted = redacted.split(secret).join('[REDACTED]');
  }
  return redacted;
};

const validateOptions = (options) => {
  validateEnvName(options.appIdEnvVar, 'app_id_env_var');
  validateEnvName(options.privateKeyEnvVar, 'private_key_env_var');
  validateEnvName(options.installationIdEnvVar, 'installation_id_env_var');
  validateEnvName(options.tokenEnvVar, 'token_env_var');
  validateEnvName(options.exportEnvPrefix, 'export_env_prefix');

  if (!INSTALLATION_LOOKUPS.has(options.installationLookup)) {
    throw new InputError('installation_lookup must be repo, org, user, app-list, or none.');
  }
};

export const createToken = async (options, { env = process.env, fetchImpl = globalThis.fetch, now = new Date() } = {}) => {
  validateOptions(options);

  const appId = clean(options.appId) || clean(env[options.appIdEnvVar]);
  if (!appId) {
    throw new InputError('GitHub App ID is required through app_id or app_id_env_var.');
  }

  const privateKey = normalizePrivateKey(env[options.privateKeyEnvVar], { base64: options.privateKeyBase64 });
  const explicitInstallationId = clean(options.installationId) || clean(env[options.installationIdEnvVar]);
  const repositories = parseCommaList(options.repositories);
  const repositoryIds = parseRepositoryIds(options.repositoryIds);
  const permissions = parsePermissions(options.permissions);

  if (repositories.length > 0 && repositoryIds.length > 0) {
    throw new InputError('repositories and repository_ids cannot be used together.');
  }

  const jwt = createAppJwt({ appId, privateKey, now });
  const installationId =
    explicitInstallationId || (await lookupInstallationId({ options, jwt, fetchImpl }));

  const tokenResponse = await requestGitHub({
    apiUrl: options.githubApiUrl,
    apiVersion: options.githubApiVersion,
    method: 'POST',
    requestPath: `/app/installations/${encodePathPart(installationId)}/access_tokens`,
    authToken: jwt,
    body: buildAccessTokenBody({ repositories, repositoryIds, permissions }),
    fetchImpl,
  });

  if (!tokenResponse.token) {
    throw new GitHubApiError('GitHub access token response did not include a token.');
  }

  const repositoryNames = Array.isArray(tokenResponse.repositories)
    ? tokenResponse.repositories.map((repository) => repository.name).filter(Boolean)
    : repositories;

  const metadata = {
    installation_id: Number.isNaN(Number(installationId)) ? installationId : Number(installationId),
    expires_at: tokenResponse.expires_at,
    repository_selection: tokenResponse.repository_selection,
    permissions: tokenResponse.permissions ?? {},
    repositories: repositoryNames,
    token_exported_to: options.tokenEnvVar,
  };

  if (options.includeTokenInOutputJson) {
    metadata.token = tokenResponse.token;
  }

  writeOutputJson({ outputJson: options.outputJson, metadata });
  writeBashEnv({
    bashEnvPath: env.BASH_ENV,
    tokenEnvVar: options.tokenEnvVar,
    exportEnvPrefix: options.exportEnvPrefix,
    outputJson: options.outputJson,
    result: {
      token: tokenResponse.token,
      installationId,
      expiresAt: tokenResponse.expires_at,
    },
  });

  return metadata;
};

const main = async () => {
  let options;
  try {
    options = readRuntimeOptions(process.env);
    const metadata = await createToken(options);
    console.log(
      `GitHub App installation token exported to ${metadata.token_exported_to}; metadata written to ${options.outputJson}.`,
    );
  } catch (error) {
    const privateKey = options?.privateKeyEnvVar ? process.env[options.privateKeyEnvVar] : undefined;
    const message = error?.stack || error?.message || String(error);
    console.error(redact(message, [privateKey]));
    process.exitCode = 1;
  }
};

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  await main();
}
